deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,556 @@
+
+    Parameters
+    ----------
+    frame : np.ndarray
+         positions of atoms
+    original_pdb : str
+         pdb file with initial condition to be used for topology
+    output_pdb_fn : str
+         where to write an outlier
+
+    """
+    pdb = PDBFile(str(original_pdb))
+    with open(str(output_pdb_fn), "w") as f:
+        PDBFile.writeFile(pdb.getTopology(), frame, f)
+
+
+def write_top_outliers(
+    cfg: OutlierDetectionConfig,
+    tmp_dir: str,
+    top: Tuple[np.ndarray, np.ndarray, np.ndarray],
+):
+    """Save to pdb files top outliers
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    tmp_dir : str
+          Temporary directory to write outliers to
+    top : Tuple[np.ndarray, np.ndarray, np.ndarray]
+          top `N` positions, velocities, md5sums where
+          `N` is equal to the number of the simulations.
+
+    """
+    positions = top[0]
+    velocities = top[1]
+    md5s = top[2]
+
+    for p, v, m in zip(positions, velocities, md5s):
+        outlier_pdb_file = f"{tmp_dir}/{m}.pdb"
+        outlier_v_file = f"{tmp_dir}/{m}.npy"
+        write_pdb_frame(p, cfg.init_pdb_file, outlier_pdb_file)
+        np.save(outlier_v_file, v)
+
+
+def write_db(top, tmp_dir):
+    """Create and save a database of outliers to be used by simulation"""
+    outlier_db_fn = f"{tmp_dir}/OutlierDB.pickle"
+    outlier_files = list(map(lambda x: f"{tmp_dir}/{x}.pdb", top[2]))
+    rmsds = top[3]
+    db = OutlierDB(tmp_dir, list(zip(rmsds, outlier_files)))
+    with open(outlier_db_fn, "wb") as f:
+        pickle.dump(db, f)
+    return db
+
+
+def publish(tmp_dir, published_dir):
+    """Publish outliers and the corresponding database for simulations to pick up"""
+    dbfn = f"{published_dir}/OutlierDB.pickle"
+    subprocess.getstatusoutput(f"touch {dbfn}")
+
+    mylock = LockFile(dbfn)
+
+    mylock.acquire()
+    print(
+        subprocess.getstatusoutput(
+            f"rm -rf {published_dir}/*.pdb {published_dir}/*.npy"
+        )
+    )
+    print(subprocess.getstatusoutput(f"mv {tmp_dir}/* {published_dir}/"))
+    mylock.release()
+
+    return
+
+
+def top_outliers(
+    cfg: OutlierDetectionConfig,
+    cvae_input: Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray],
+    outlier_list: np.ndarray,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Find top num_sim outliers sorted by rmsd.
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    cvae_input : Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+            steps, positions, velocities, md5sums, rmsds
+    outlier_list : np.ndarray
+            indices corresponding to outliers
+
+    Returns
+    -------
+    Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+         Positions, velocities, md5sums, rmsds, outlier
+         indices of outliers, sorted in ascending order by rmsd
+
+    """
+    N = cfg.num_sim
```

Comment:
I don't see `num_sim` in the OutlierDetectionConfig and my IDE says it doesn't exist. It looks like it is defined in the test suite in TASK_CONFIG_AGENT, so it should be added to OutlierDetectionConfig as well.

Reply:
Done